src/NLog.Database/DatabaseParameterInfo.cs

```diff
@@ -203,61 +238,15 @@ internal bool SetDbType(IDbDataParameter dbParameter)
             return true;    // DbType not in use
         }
 
-        private static Type TryParseDbType(string dbTypeString)
+        private static Type TryParseDbType(string dbTypeName)
         {
-            string[] dbTypeNames = dbTypeString?.Split(new[] { '.' }, StringSplitOptions.RemoveEmptyEntries);
-            dbTypeString = dbTypeNames?.Length > 0 ? dbTypeNames[dbTypeNames.Length - 1] : dbTypeString;
+            // retrieve the type name if a full name is given
+            dbTypeName = dbTypeName?.Substring(dbTypeName.LastIndexOf('.') + 1);
 
-            if (string.IsNullOrEmpty(dbTypeString))
+            if (string.IsNullOrEmpty(dbTypeName))
                 return null;
-            if (dbTypeString.IndexOf("Bool", StringComparison.OrdinalIgnoreCase) >= 0)
-                return typeof(bool);
-            if (dbTypeString.IndexOf("Timestamp", StringComparison.OrdinalIgnoreCase) >= 0)
-                return typeof(DateTime);
-            if (dbTypeString.IndexOf(nameof(System.Data.DbType.DateTimeOffset), StringComparison.OrdinalIgnoreCase) >= 0)
-                return typeof(DateTimeOffset);
-            if (dbTypeString.IndexOf(nameof(System.Data.DbType.DateTime), StringComparison.OrdinalIgnoreCase) >= 0)
-                return typeof(DateTime);
-            if (dbTypeString.IndexOf(nameof(System.Data.DbType.Date), StringComparison.OrdinalIgnoreCase) >= 0)
```

Comment:
Notice how this matches, when type-string actual only contains the word "Date". This is no longer the case.

Reply:
Indeed.

It was to handle (at least) the `DbType.DateTime2` case; it is now an entry of `_typesByDbTypeName`.  
The same apply for `String` variations (ie: `DbType.AnsiString`, `DbType.AnsiStringFixedLength` and `DbType.StringFixedLength`).

To ensure full coverage, it is necessary to know all possible values of `dbTypeString` that need to be handled. 
[The documentation](https://github.com/NLog/NLog/wiki/Database-target#sql-statement) refers to _members_ of the [`DbType`](https://docs.microsoft.com/en-us/dotnet/api/system.data.dbtype), [`SqlDbType`](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqldbtype) and [`NpgsqlDbType`](https://www.npgsql.org/doc/api/NpgsqlTypes.NpgsqlDbType.html) enums.

Current State
---

For now, are handled:
  - Most of the members from `DbType`.
    - Those values are the ones tested.
    - `DbType.Binary` is not handled.
  - Some members from `SqlDbType` (witch fortuitously match `DbType` names) plus `Timestamp`.
    - Some basic values are not handle (like `Bit`, `UniqueIdentifier`, `Text`, ...)
  - Some members from `NpgsqlDbType` (witch fortuitously match `DbType` names) plus `Timestamp`.
    - Some basic values are not handle (like `Text`, `Numeric`, `Uuid`, ...)
    - Some values are badly interpreted (like `TimestampRange` interpreted as `Timestamp`)


With this PR
---

 Are handled:
  - All the members from `DbType`
  - Some members from [`SqlDbType`](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqldbtype) (witch fortuitously match `DbType` names) plus `Timestamp`.
    - Some basic values are not handle (like `Bit`, `UniqueIdentifier`, `Text`, ...)
  - Some values from [`NpgsqlDbType`](https://www.npgsql.org/doc/api/NpgsqlTypes.NpgsqlDbType.html) (witch fortuitously match `DbType` names) plus `Timestamp`.
    - Some basic values are not handle (like `Text`, `Numeric`, `Uuid`, ...)
    - `TimestampRange` is no more badly interpreted as `Timestamp`

Reply:
Not sure I understand doesn't `DateTime` + `Date` + `DateTime2` map to `typeof(DateTime)` ?

Reply:
> Not sure I understand doesn't `DateTime`, `Date` and `DateTime2` map to `typeof(DateTime)` ?

Yes.

> Notice the current implementation is just best-effort to make a guess independent of DbProvider. Details like "Timestamp" for SqlDbType-provide might not be the same for a different DbProvider.

`Timestamp` meaning depends on the DbProvider. Actual implementation make it a `DateTime` choosing _NpgsqlDb_ over _SqlDb_.

I think that should not be the case and `Timestamp` should not be mapped.

Members of `System.Data.DbType` are representative: they map to any useful type.